Fixes #2725: Bugform validation improvements

ksy36 opened the pull request:
r? @miketaylr

miketaylr commented on webcompat/static/js/lib/bugform-validation.js:
```diff
@@ -0,0 +1,48 @@
+  var GITHUB_REGEXP = /^[a-z\d](?:[a-z\d]|-(?=[a-z\d])){0,38}$/i;
+  var ALLOWED_IMG_FORMAT = ["jpg", "jpeg", "jpe", "png", "gif", "bmp"];
+
+  var isReportableURL = function(url) {
```
Note: I guess this validation kinda sucks, since it will accept just `http:blah`. But we can improve that in a follow-up bug. Historically we haven't wanted to be *too* strict, but we can probably improve things here.

ksy36 replied:
yeah I've noticed that :) wasn't sure whether I should change it to be more strict. I'll create a bug for that

miketaylr commented on webcompat/static/js/lib/bugform.js:
```diff
@@ -285,99 +293,106 @@ BugForm.prototype.enableSubmits = function() {
 
+/* determines function based on whether validation returns true or false */
+BugForm.prototype.determineValidityFunction = function(func, field, silent) {
+  if (func(field)) return "makeValid";
```
nit: can we do multi-line conditional here? (or did prettier do this for you?)

ksy36 replied:
added

miketaylr commented on webcompat/static/js/lib/bugform.js:
```diff
@@ -285,99 +293,106 @@ BugForm.prototype.enableSubmits = function() {
   // null out input.value so we get a consistent
   // change event across browsers
-  event.target.value = null;
+  if (event) event.target.value = null;
```
nit: can we do a multi-line conditional here?

miketaylr commented on webcompat/static/js/lib/bugform.js:
```diff
@@ -410,88 +421,105 @@ BugForm.prototype.checkForm = function() {
+  this.showError(id);
+  // someone might decide to just not select an image after seeing the error,
+  // so buttons shouldn't be disabled
+  if (id !== "image") this.disableSubmits();
```
nit: multi-line conditional.

ksy36 commented on webcompat/templates/layout.html:
```diff
@@ -32,6 +32,7 @@
 <script src="{{ url_for('static', filename='js/lib/flash-message.js') }}"></script>
 <script src="{{ url_for('static', filename='js/lib/homepage.js') }}"></script>
 <script src="{{ url_for('static', filename='js/lib/autogrow-textfield.js') }}"></script>
+<script src="{{ url_for('static', filename='js/lib/bugform-validation.js') }}"></script>
```
thanks, added that

ksy36 commented on webcompat/templates/home-page/form.html:
```diff
@@ -12,7 +12,7 @@ <h1 class="headline-1">Report Site Issue</h1>
   <div class="grid-cell">
     <div class="form-text form-element js-Form-group">
       <div class="js-Form-information form-label-message">
-        {{ form.url.label(class_='form-label form-required') }}
```
good catch, I've missed that when moving changes from v1 branch.. there is a `(mandatory)` as well for two more fields in the `form.py`

ksy36 commented on webcompat/templates/home-page/form.html:
```diff
@@ -26,7 +26,7 @@ <h1 class="headline-1">Report Site Issue</h1>
     <div class="form-radio form-element js-Form-group">
       <div class="js-Form-information form-label-message">
-        <label class="form-label form-required" for="problem_category">Type of bug</label>
+        <label class="form-label" for="problem_category">Type of bug (mandatory)</label>
```
fixed that, there are two more :) but (mandatory) is being added on the python side

ksy36 commented:
Yeah, sorry :) I had individual commits in v1 branch, but then after we refactored the form [here](https://github.com/webcompat/webcompat.com/pull/2885), decided to move these validation changes to v2 branch and base them on the refactored version. So they all landed in one commit..

For the required fields there were 2 more missing `(mandatory)` , added them.

ksy36 commented:
Actually I've just noticed a small issue, will be fixing it shortly.

miketaylr commented:
> Yeah, sorry :)

No issue! Just makes it a bit easier to review in chunks.

ksy36 commented:
Ok the last issue is fixed now. I made it so the errors don't show up on page refresh or when coming from the report site issue extension, so a user will see them only when clicking on submit.

miketaylr commented:
@karlcow can we push this branch to staging to do a quick sanity check the form still behaves as we expect. It LGTM locally.

miketaylr commented:
(we can merge now, or wait until we check it out on staging, no strong preference here)

karlcow commented:
@ksy36 @miketaylr this branch is deployed on staging.

miketaylr commented:
> @ksy36 @miketaylr this branch is deployed on staging.

Thanks Karl. I'll poke around before merging.

miketaylr commented:
@ksy36 I think I discovered a bug:

https://staging.webcompat.com/issues/1882

STR:

open the bug form on staging.webcompat.com/issues/new
put in a URL
select a problem type
write some summary
click back to the URL input
select all the text (cmd+a, on mac, for example)
delete all the text (cmd+x, on mac, for example)
without changing focus, use the mouse to click submit.

Expected: the form should not submit
Actual: the form submits, with a None URL.

I checked the STR on production, and if you delete all the text, we run validation. So this would technically be a regression (probably a super-corner case). 

Do you want to try to fix here? Or in a follow up bug?

ksy36 commented:
@miketaylr  weird, I'm not able to reproduce - the button becomes slightly grey and disabled once I press `cmd+x` and the text clears out. When I click on the button, the url error shows up as expected. 
Am I missing something?

miketaylr commented:
I'm not 100% sure. But it's pretty edge-casey, so let's merge here and I'll file a follow up if I can reproduce in a new profile.
